src/Parser.hs

```diff
@@ -5,113 +5,112 @@ module Parser
-import Text.Parsec.String (Parser)
-import Text.ParserCombinators.Parsec.Combinator (many1)
+import Text.Parsec
+    ( choice
+    , option
+    , optionMaybe
+    , (<|>)
+    , (<?>)
+    , parse
+    , many
+    , many1
+    , unexpected
+    , eof )
+import Text.Parsec.String (Parser) -- TODO: We're gonna want to do better than
+                                   --       this... Fortunately that won't
+                                   --       break anything, if we're careful.
 import Text.Parsec.Char (string, char, oneOf, digit)
 import Text.Parsec.Prim (try)
 
 import Data.Maybe (fromMaybe)
 
+-- TODO: Let's be a little more robust here
 parseChord :: String -> Maybe Chord
-parseChord s =
-  case parse parserChord "" s of
-    Left err  -> Nothing
-    Right xs  -> Just xs
-
+parseChord s = rightToMaybe $ parse parserChord "" s
 
+-- N.B. This should only be used when it's absolutely necessary
 (<||>) :: Parser a -> Parser a -> Parser a
 p <||> q = try p <|> q
 
 parserAccidental :: Parser Accidental
 parserAccidental =
-  option AccNatural $ parser '#' AccSharp <||> parser 'b' AccFlat
+  option AccNatural (parseAllOf '#' <|> parseAllOf 'b')
   where
-    parser :: Char -> (Int -> Accidental) -> Parser Accidental
-    parser symbol constructor =
-      do
-        cs <- many1 $ char symbol
-        return $ constructor (length cs)
+    parseAllOf :: Char -> Parser Accidental
+    parseAllOf acc = do let constructor = case acc of
+                                            '#' -> AccSharp
+                                            'b' -> AccFlat
+                        accs <- many1 $ char acc
+                        return $ constructor $ length accs
 
 parserRoot :: Parser Root
-parserRoot =
-  do
-    noteChar <- oneOf "ABCDEFG"
-    let note = read [noteChar]
-    Root note <$> parserAccidental
+parserRoot = do noteChar <- oneOf "ABCDEFG"
+                let note = read [noteChar]
+                Root note <$> parserAccidental
 
 parserQuality :: Parser Quality
-parserQuality =
-       parser "^" QMajor
-  <||> parser "M" QMajor
-  <||> parser "-" QMinor
-  <||> parser "m" QMinor
-  <||> parser "dim" QDiminished
-  <||> parser "o" QDiminished
-  <||> parser "aug" QAugmented
-  <||> parser "+" QAugmented
+parserQuality = choice $ parseQualfromString <$> qualStrings
   where
-    parser :: String -> Quality -> Parser Quality
-    parser literal quality =
-      do
-        _ <- string literal
-        return quality
+    qualStrings = [ ("^",   QMajor)
+                  , ("M",   QMajor)
+                  , ("-",   QMinor)
+                  , ("m",   QMinor)
+                  , ("dim", QDiminished)
+                  , ("o",   QDiminished)
+                  , ("aug", QAugmented)
+                  , ("+",   QAugmented )
+                  ]
+
+    parseQualfromString :: (String, Quality) -> Parser Quality
+    parseQualfromString (qualName, quality) = string qualName >> return quality
```

Comment:
This was another fun refactor, but it warrants some explaining.

First, notice the redundancy in the original code (around the uses of `<||>`); the function call appears on every line. This led to the first refactor: Give a name to a collection of all the `String` / `Quality` associations. 

Next, we use the `choice` combinator from Parsec, which takes a list of parsers and tries each one in order until a success; this is a nice abstraction over the pattern in the original code.

Finally, the helper function was renamed and dramatically simplified. This simplification is a natural result of de-sugaring the original `do`-block, since we are performing a parsing action, ignoring its result, and immediately performing another parsing action (the `return quality` action.) This is written most simply using the `>>` combinator for monads, which is defined as `a >> b = a >>= (\_ -> b)`.